pkg/collector/listeners/kubelet.go

```diff
@@ -0,0 +1,233 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017 Datadog, Inc.
+
+// +build kubelet
+
+package listeners
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/tagger"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/kubelet"
+	log "github.com/cihub/seelog"
+)
+
+// KubeletListener listen to kubelet pod creation
+type KubeletListener struct {
+	watcher    *kubelet.PodWatcher
+	services   map[ID]Service
+	newService chan<- Service
+	delService chan<- Service
+	ticker     *time.Ticker
+	stop       chan bool
+	m          sync.RWMutex
+}
+
+// PodContainerService implements and store results from the Service interface for the Kubelet listener
+type PodContainerService struct {
+	ID            ID
+	ADIdentifiers []string
+	Hosts         map[string]string
+	Ports         []int
+	Pid           int
+	PodInfos      *kubelet.Pod
+}
+
+func init() {
+	Register("kubelet", NewKubeletListener)
+}
+
+func NewKubeletListener() (ServiceListener, error) {
+	watcher, err := kubelet.NewPodWatcher(5 * time.Second)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to kubelet, Kubernetes listener will not work: %s", err)
+	}
+	return &KubeletListener{
+		watcher:  watcher,
+		services: make(map[ID]Service),
+		ticker:   time.NewTicker(5 * time.Second),
+		stop:     make(chan bool),
+	}, nil
+}
+
+func (l *KubeletListener) Listen(newSvc chan<- Service, delSvc chan<- Service) {
+	// setup the I/O channels
+	l.newService = newSvc
+	l.delService = delSvc
+
+	go func() {
+		for {
+			select {
+			case <-l.stop:
+				return
+			case <-l.ticker.C:
+				// Compute new/updated pods
+				updatedPods, err := l.watcher.PullChanges()
+				if err != nil {
+					log.Error(err)
+					continue
+				}
+				for _, pod := range updatedPods {
+					// Ignore pending/failed/succeeded/unknown states
+					if pod.Status.Phase == "Running" {
+						l.processNewPod(pod)
+					}
+				}
+				// Compute deleted pods
+				expiredContainerList, err := l.watcher.ExpireContainers()
+				if err != nil {
+					log.Error(err)
+					continue
+				}
+				for _, containerID := range expiredContainerList {
+					l.removeService(ID(containerID))
+				}
+			}
+		}
+	}()
+}
+
+func (l *KubeletListener) Stop() {
+	l.ticker.Stop()
+	l.stop <- true
+}
+
+func (l *KubeletListener) processNewPod(pod *kubelet.Pod) {
+	for _, container := range pod.Status.Containers {
+		l.createService(ID(container.ID), pod)
+	}
+}
+
+func (l *KubeletListener) createService(id ID, pod *kubelet.Pod) {
+	svc := PodContainerService{
+		ID:       id,
+		PodInfos: pod,
+	}
+
+	podName := pod.Metadata.Name
+	_, err := svc.GetADIdentifiers()
```

Comment:
All this caching + pre-compute logic was relevant for the DockerListener (as the info was not available in the `docker ps` on start, but I think that's not needed for the Kubelet, as we have all the info when creating the service.
Let's just set `svc.Hosts` and friends there and make the `Get*()` basic getters